#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [2]:
df=pd.read_csv('../data/supermarket_sales.csv')

In [4]:
df.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4  
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [5]:
df.groupby('Branch')['gross income'].sum()

Branch
A    5057.1605
B    5057.0320
C    5265.1765
Name: gross income, dtype: float64

Філіал **'C'** має найвищий сумарний прибуток - **5265**.

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [15]:
df.groupby(['Branch', 'City'])['gross income'].sum()

Branch  City     
A       Yangon       5057.1605
B       Mandalay     5057.0320
C       Naypyitaw    5265.1765
Name: gross income, dtype: float64

Отже, в кожному місті знаходиться по одній філії. В **Naypyitaw** знаходиться філія **C**, яка має найвищий прибуток, отже саме там варто розглянути відкриття ще одного магазину.

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [41]:
invoices_by_category = df.pivot_table( index='Branch', columns='Product line', values='Invoice ID', aggfunc='count')

In [42]:
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [81]:
invoices_by_category['Electronic accessories'].idxmax()

'A'

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


1. **Чи жінки роблять більше покупок за чоловіків та чи на  більший середній чек?**

In [118]:
df.groupby('Gender').Total.agg(['count', 'sum', 'mean']).round(2)

count        sum    mean
Gender                          
Female    501  167882.92  335.10
Male      499  155083.82  310.79

**Кількість покупок** серед чоловіків і жінок майже **однакова (501 проти 499)**. Проте, **жінки** роблять дорожчі покупки: **середня сума чеку** становить **335.10**, тоді як у **чоловіків — 310.79**.

**Рекомендується** зосередити маркетингові заходи на жіночій аудиторії, оскільки вона демонструє **вищу середню суму покупки**. Доцільно проаналізувати категорії товарів, у яких жінки витрачають найбільше — це допоможе оптимізувати асортимент і рекламу. Для чоловіків рекомендовано запровадити тематичні акції або бонусні пропозиції, щоб збільшити середній чек.

2. **Чи є залежність між розміром покупки (сума чеку) та рівнем задоволеності клієнтів (рейтинг).**

In [129]:
df['Rating_round'] = df['Rating'].round(0)

In [168]:
df.groupby('Rating_round').Total.agg(['count', 'mean', 'sum']).round(2).rename(columns={
      'count': 'Count',
      'mean': 'Average',
      'sum': 'Total_Sum'
  })

Count  Average  Total_Sum
Rating_round                           
4.0             102   368.10   37546.71
5.0             133   341.56   45427.16
6.0             193   305.36   58934.84
7.0             159   310.83   49421.42
8.0             179   311.22   55708.58
9.0             141   339.23   47831.28
10.0             93   302.12   28096.76

**Найбільше покупок** припадає на середні оцінки **(6–8 балів)**. При цьому **найвищий середній чек** спостерігається саме серед клієнтів із **нижчими рейтингами (4–5 балів)**, тоді як покупці з **високими оцінками (9–10) витрачають менше.

**Рекомендації:**

 - Звернути особливу увагу на покупців із великими чеками та низькими рейтингами — вони приносять значний дохід, але менш задоволені.

- Проаналізувати причини їхніх оцінок і покращити сервіс або якість обслуговування для цієї групи клієнтів.

Перевіримо чи рівень **задоволеності** відрізняється між філіалами **(Branch)**.

In [177]:
df.groupby('Branch').Rating.mean().round(2)

Branch
A    7.03
B    6.82
C    7.07
Name: Rating, dtype: float64

**Висновок:**
Різниця між філіями невелика, але **філія B** може потребувати додаткової уваги — її клієнти оцінюють покупки **нижче**.

**Рекомендації:**
- Перевірити якість обслуговування, асортимент та швидкість обслуговування у **філії B**.
- Впровадити успішні практики з філій A і C.

Перевіримо чи **рівень задоволеності** залежить від категорії товару **(Product line)**

In [176]:
df.groupby('Product line').Rating.mean().round(2).sort_values(ascending=False)

Product line
Food and beverages        7.11
Fashion accessories       7.03
Health and beauty         7.00
Electronic accessories    6.92
Sports and travel         6.92
Home and lifestyle        6.84
Name: Rating, dtype: float64

Категорії товарів отримали досить **схожі рейтинги**, що свідчить про **стабільний рівень задоволеності клієнтів**. **Найвищі оцінки** мають **Food and beverages та Fashion accessories**, тоді як **найнижчі — Home and lifestyle**.

**Рекомендується** покращити якість та консультації в категоріях Home and lifestyle, де рейтинг нижчий. Категорії з високими оцінками варто підтримувати на поточному рівні.

3. **Клієнти, які є учасниками програми лояльності (Member), здійснюють дорожчі покупки (приносять більший дохід), ніж звичайні клієнти (Normal)?**

In [181]:
df.groupby('Customer type').Total.agg(['count', 'mean', 'sum', 'min', 'max']).round(2)

count    mean        sum    min      max
Customer type                                          
Member           501  327.79  164223.44  10.68  1042.65
Normal           499  318.12  158743.30  12.69  1039.29

**Висновки:**

- **Середній чек** в учасників програми **(327.79) вищий**, ніж у звичайних клієнтів **(318.12)**.

- **Загальна сума** покупок також **більша у групи Member**, хоча кількість покупок майже однакова.

- **Програма лояльності** вже дає помітний, хоч і невеликий **позитивний ефект**.

**Рекомендації:**

- **Розвивати** програму лояльності, оскільки вона стимулює вищі витрати.

- Заохочувати звичайних клієнтів **оформлювати членство** (бонуси за першу покупку, промокоди).

- Використати дані про членів для **персоналізованих пропозицій**, щоб ще збільшити середній чек.